### ISSUE TITLE
Confused about type signatures when introducing `Error` effect when interpreting

### ISSUE THREAD

**Issue by oberblastmeister** [haskell-effectful/effectful#59]
I am not sure how to best explain this, but the docs show in the file systems example that you can reinterpret the `FileSystem` effect with an internal `State` effect and an `Error` effect that has a constraint. However, the type signature is this
```haskell
runFilesystemPure ::
  Error FsError :> es =>
  M.Map FilePath String ->
  Eff (Filesystem : es) a ->
  Eff es a
```
The `Error FsError` has a constraint on both the entry and exit `Eff`. So if we define a function
```haskell
runFileSystemPure' :: M.Map FilePath String -> Eff (FileSystem ': Error FsError ': es) a -> Eff es (Either (CallStack, FsError) a)
runFileSystemPure' fs = runFileSystemPure fs >>> runError @FsError
```
Why is the `Error FsError` effect there? Because it is only introduced when the function is interpreted because you cannot possibly throw a `FsError` other than inside of the interpreter. Sorry if this is dumb on my part, but I am a little confused.

**Comment by mmhat** [comment]
I am not sure if I understand your question but I will try to answer what I assume you meant.

The `reinterpret` function allows you to interpret the effect in a local environment prepended to the parent "environment" stack:
The (simplified) type signature of `reinterpret` looks like this:
```
reinterpret :: (Eff handlerEs a -> Eff es b) -> EffectHandler e handlerEs -> Eff (e : es) a -> Eff es b
```
In the example we pass `evalState fs0` as the first argument to that function (See [here](https://github.com/arybczak/effectful/blob/4ab5e7b5f61c56b5243396358984353565c781e9/effectful/examples/FileSystem.hs#L60)) where `fs0 :: Map FilePath String`. Then:
```
evalState fs0 :: Eff (State (M.Map FilePath String) : es) a -> Eff es a
reinterpret (evalState fs0) :: EffectHandler e (State (Map FilePath String) : es) -> Eff (e : es) a -> Eff es a
```
Since we use `throwError . FsError` in the `EffectHandler` a constraint `Error FsError :> es` is imposed on our effect row `es`.

**TLDR;** We reinterpret the `FileSystem` effect as `State`, but not as `Error`. We do however use the `Error` effect and there must be a handler for that effect in the surrounding environment. If we were reinterpreting the filesystem effect in terms of both `State` and `Error` then our call to `reinterpet` would be something like `reinterpret (runError @FsError . evalState fs0)`.

The [documentation](https://github.com/arybczak/effectful/blob/4ab5e7b5f61c56b5243396358984353565c781e9/effectful-core/src/Effectful/Internal/Env.hs#L55) of the implementation of the `Env` type used to store the environment might be of interest here, too.

**Comment by arybczak** [comment]
The above seems to answer the question :+1:
